File: 2022/task1/task1.py
def two_stacks(max_sum, a, b):
    a = list(reversed(a))
    b = list(reversed(b))
    total_sum = 0
    len_a = len(a)
    len_b = len(b)

    selected_numbers = []
    # берем все элементы из первого стека пока не превысим лимит
    for _ in range(len_a):
        value = a.pop()
        if total_sum + value > max_sum:
            break
        total_sum += value
        selected_numbers.append(value)

    # будем заменять по одному из второго стека
    max_count = len(selected_numbers)
    current_count = max_count
    while len_b:
        if total_sum + b[-1] <= max_sum:
            total_sum += b.pop()
            len_b -= 1
            current_count += 1
            if current_count > max_count:
                max_count = current_count
            continue
        if not len(selected_numbers):
            break
        value_from_first_stack = selected_numbers.pop()
        total_sum -= value_from_first_stack
        current_count -= 1
    return max_count


if __name__ == '__main__':
    task_input = input().rstrip().split(' ')
    n = int(task_input[0])
    m = int(task_input[1])
    max_sum = int(task_input[2])
    a = []
    b = []

    for i in range(max(n, m)):
        [ai, bi] = input().rstrip().split(' ')
        if ai.isnumeric():
            a.append(int(ai))
        if bi.isnumeric():
            b.append(int(bi))

    result = two_stacks(max_sum, a, b)

    print(str(result))

File: 2022/task1/task1.js
const readline = require('readline').createInterface(process.stdin, process.stdout);

let aSize = null;
let bSize = null;
let maxSum = 0;
let a = [];
let b = [];
let linesCount = 0;

function twoStacks(maxSum, a, b) {
    a.reverse();
    b.reverse();
    let totalSum = 0;
    let lenA = a.length;
    let lenB = b.length;

    let selectedNumbers = [];
    // берем все элементы из первого стека пока не превысим лимит
    for (const _ of [...Array(lenA)]) {
        const value = a.pop();
        if (totalSum + value > maxSum) {
            break;
        }
        totalSum += value;
        selectedNumbers.push(value);
    }

    // будем заменять по одному из второго стека
    let maxCount = selectedNumbers.length;
    let currentCount = maxCount;
    while (lenB > 0) {
        if (totalSum + b[lenB - 1] <= maxSum) {
            totalSum += b.pop();
            lenB -= 1;
            currentCount += 1;
            if (currentCount > maxCount) {
                maxCount = currentCount;
            }
            continue;
        }
        if (selectedNumbers.length === 0) {
            break;
        }
        const valueFromFirstStack = selectedNumbers.pop()
        totalSum -= valueFromFirstStack
        currentCount -= 1
    }

    return maxCount
}

readline.on('line', (line) => {
    if (linesCount === 0) {
        [aSize, bSize, maxSum] = line.split(' ').map(item => Number(item));
    } else {
        const [ai, bi] = line.split(' ');
        if (ai !== '-') {
            a.push(Number(ai));
        }

        if (bi !== '-') {
            b.push(Number(bi))
        }
    }
    if (linesCount++ === Math.max(aSize, bSize)) {
        console.log(twoStacks(maxSum, a, b));
        readline.close();
    }
}).on('close', () => process.exit(0));
